CDI Test failures with GlassFish 7

Running the CDI TCK (using the runner in GlassFish) there are 11 failures as of today:

[ERROR] Failures: 

- [x]   EjbInjectionTest>Arquillian.run:138->testPassivationOfEjbs:70 » EJB java.rmi.RemoteException: CORBA BAD_OPERATION 0 No; nested exception is: 
	org.omg.CORBA.BAD_OPERATION: The delegate has not been set!  vmcid: 0x0  minor code: 0  completed: No
- [x]    AmbiguousInjectionIntoNonContextualComponentTest>Arquillian.arquillianBeforeClass:99 » Runtime Expected exception of type jakarta.enterprise.inject.spi.DeploymentException during deployment of _DEFAULT_, but no exception was thrown.
- [x]   UnsatisfiedInjectionIntoNonContextualComponentTest>Arquillian.arquillianBeforeClass:99 » Runtime Expected exception of type jakarta.enterprise.inject.spi.DeploymentException during deployment of _DEFAULT_, but no exception was thrown.
- [x]    InjectionPointTest>Arquillian.arquillianBeforeClass:99 » Runtime Expected exception of type jakarta.enterprise.inject.spi.DefinitionException during deployment of _DEFAULT_, but no exception was thrown.
- [x]    CDIProviderInitTest>Arquillian.run:138->testAccessingBeanManager:103 expected [0] but found [1]
- [x]    CDIProviderRuntimeTest>Arquillian.run:138->testGetBeanManager:76 expected [1] but found [0]
- [x]    CDIProviderRuntimeTest>Arquillian.run:138->testLookup:90 » UnsatisfiedResolution WELD-001334: Unsatisfied dependencies for type Bravo with qualifiers  

Fixed by Weld update:

- [x]   ContainerLifeCycleEventRuntimeInvocationTest>Arquillian.run:138->testAfterBeanDiscoveryEventFails:224 » NullPointer
- [x]   ContainerLifeCycleEventRuntimeInvocationTest>Arquillian.run:138->testAfterBeanDiscoveryEventFails:224 » NullPointer

See https://github.com/jakartaee/cdi/issues/624#issuecomment-1149357201

Fixed by CDI TCK update:

- [x]   PersistenceContextInjectionTest>Arquillian.run:138->testBeanTypesAndBindingTypesOfPersistenceContext:105
- [x]   PersistenceContextInjectionTest>Arquillian.run:138->testBeanTypesOfPersistenceUnit:116

See https://github.com/jakartaee/cdi-tck/issues/380




```
[ERROR] Tests run: 1837, Failures: 11, Errors: 0, Skipped: 6
```

Tests can be run from `[glassfish]/appserver/tests/tck/cdi` using:

`mvn clean install`

Individual tests can be run via

`mvn clean install -Dit.test=ExcludeFiltersTest#testExcludeSystemPropertyActivator`

Tests can be debugged via:

`mvn clean install -Dit.test=ExcludeFiltersTest#testExcludeSystemPropertyActivator -Dglassfish.suspend`

--- a/appserver/web/web-core/src/main/java/org/apache/catalina/Wrapper.java
+++ b/appserver/web/web-core/src/main/java/org/apache/catalina/Wrapper.java
@@ -341,6 +341,8 @@ public interface Wrapper extends Container {
      */
     public void load() throws ServletException;
 
+    public void tryLoad() throws ServletException;
+
 
     /**
      * Remove the specified initialization parameter from this servlet.

--- a/appserver/web/web-core/src/main/java/org/apache/catalina/core/StandardContext.java
+++ b/appserver/web/web-core/src/main/java/org/apache/catalina/core/StandardContext.java
@@ -4977,37 +4977,58 @@ public class StandardContext extends ContainerBase implements Context, ServletCo
      */
     public void loadOnStartup(Container children[]) throws LifecycleException {
         // Collect "load on startup" servlets that need to be initialized
-        Map<Integer, List<Wrapper>> map = new TreeMap<>();
+        Map<Integer, List<Wrapper>> loadOnStartupServlets = new TreeMap<>();
+        List<Wrapper> nonLoadOnStartupServlets = new ArrayList<>();
+
         for (Container aChildren : children) {
             Wrapper wrapper = (Wrapper) aChildren;
+
             int loadOnStartup = wrapper.getLoadOnStartup();
             if (loadOnStartup < 0) {
-                continue;
+                nonLoadOnStartupServlets.add(wrapper);
+            } else {
+                loadOnStartupServlets.computeIfAbsent(loadOnStartup, e -> new ArrayList<>())
+                                     .add(wrapper);
             }
+        }
 
-            Integer key = loadOnStartup;
-            List<Wrapper> list = map.get(key);
-            if (list == null) {
-                list = new ArrayList<>();
-                map.put(key, list);
+        // Combine the load on startup and non load on startup in one list, with the
+        // latter loading after the ones with an explicit priority (load level).
+        List<Wrapper> allServlets = new ArrayList<>();
+        for (List<Wrapper> samePriorityServlets : loadOnStartupServlets.values()) {
+            for (Wrapper wrapper : samePriorityServlets) {
+                allServlets.add(wrapper);
             }
-            list.add(wrapper);
         }
 
         // Load the collected "load on startup" servlets
-        for (List<Wrapper> list : map.values()) {
-            for (Wrapper wrapper : list) {
-                try {
-                    wrapper.load();
-                } catch (ServletException e) {
-                    getServletContext().log(
-                        format(rb.getString(SERVLET_LOAD_EXCEPTION), neutralizeForLog(getName())),
-                        StandardWrapper.getRootCause(e));
+        for (Wrapper wrapper : allServlets) {
+            try {
+                wrapper.load();
+            } catch (ServletException e) {
+                getServletContext().log(
+                    format(rb.getString(SERVLET_LOAD_EXCEPTION), neutralizeForLog(getName())),
+                    StandardWrapper.getRootCause(e));
 
                     // NOTE: load errors (including a servlet that throws
                     // UnavailableException from the init() method) are NOT
                     // fatal to application startup
                     throw new LifecycleException(StandardWrapper.getRootCause(e));
+            }
+        }
+
+        if (Boolean.parseBoolean(System.getProperty("glassfish.try.preload.servlets", "true"))) {
+            // Also load the other servlets, which is one way to pass the CDI TCK, specifically
+            // ContainerEventTest#testProcessInjectionTargetEventFiredForServlet and adhere to the rule
+            // that injection points for Servlets have to be processed during start.
+            for (Wrapper wrapper : nonLoadOnStartupServlets) {
+                try {
+                    wrapper.tryLoad();
+                } catch (Throwable t) {
+                    // Only log errors, don't fail anything.
+                    getServletContext().log(
+                            format(rb.getString(SERVLET_LOAD_EXCEPTION), neutralizeForLog(getName())),
+                            t);
                 }
             }
         }

--- a/appserver/web/web-core/src/main/java/org/apache/catalina/core/StandardWrapper.java
+++ b/appserver/web/web-core/src/main/java/org/apache/catalina/core/StandardWrapper.java
@@ -1114,6 +1114,23 @@ public class StandardWrapper extends ContainerBase implements ServletConfig, Wra
         initServlet(instance);
     }
 
+    @Override
+    public synchronized void tryLoad() throws ServletException {
+        try {
+            instance = loadServlet();
+            initServlet(instance);
+        } catch (Throwable t) {
+            instance = null;
+            available = 0l;
+            classLoadTime = 0;
+            loadTime = 0l;
+            instanceInitialized = false;
+
+            throw t;
+        }
+    }
+
+
     /**
      * Creates an instance of the servlet, if there is not already at least one initialized instance.
      */

